storage/aptosdb/src/db_debugger/truncation/mod.rs

```diff
@@ -0,0 +1,311 @@
+use aptos_schemadb::{
+    schema::{Schema, SeekKeyCodec},
+    ReadOptions, SchemaBatch, DB,
+};
+use aptos_types::transaction::Version;
+use clap::Parser;
+use std::{path::PathBuf, sync::Arc};
+
+#[derive(Parser)]
+#[clap(about = "Delete all data after the provided version.")]
+pub struct Cmd {
+    #[clap(long, parse(from_os_str))]
+    db_dir: PathBuf,
+
+    target_version: u64,
+
+    ledger_db_batch_size: usize,
+}
+
+impl Cmd {
+    pub fn run(self) -> Result<()> {
+        let (ledger_db, state_merkle_db, _kv_db) = AptosDB::open_dbs(
+            &self.db_dir,
+            RocksdbConfigs::default(),
+            /*readonly=*/ false,
+        )?;
+
+        // TODO(grao): Handle kv db once we enable it.
+
+        let ledger_db_version = Self::get_current_version_in_ledger_db(&ledger_db)?
+            .expect("Current version of ledger db must exist.");
+        let state_merkle_db_version =
+            Self::get_current_version_in_state_merkle_db(&state_merkle_db)?
+                .expect("Current version of state merkle db must exist.");
+
+        assert!(self.target_version <= ledger_db_version);
+        assert!(ledger_db_version >= state_merkle_db_version);
+
+        // TODO(grao): Ensure target version is not pruned.
+
+        let state_merkle_target_version =
+            Self::find_tree_root_at_or_before(&state_merkle_db, self.target_version)?.expect(
+                &format!(
+                    "Could not find a valid root before or at version {}, maybe it was pruned?",
+                    self.target_version
+                ),
+            );
+
+        Self::truncate_state_merkle_db(state_merkle_db, state_merkle_target_version)?;
+
+        Self::truncate_ledger_db(
+            ledger_db,
+            ledger_db_version,
+            self.target_version,
+            self.ledger_db_batch_size,
+        )?;
+
+        if state_merkle_target_version < self.target_version {
+            // TODO(grao): If state merkle db is not consistent with ledger db, replay the write set to
+            // make them consistent.
+        }
+
+        Ok(())
+    }
+
+    fn get_current_version_in_ledger_db(ledger_db: &DB) -> Result<Option<Version>> {
+        let mut iter = ledger_db.iter::<TransactionInfoSchema>(ReadOptions::default())?;
+        iter.seek_to_last();
+        Ok(iter.next().transpose()?.map(|item| item.0))
+    }
+
+    fn get_current_version_in_state_merkle_db(state_merkle_db: &DB) -> Result<Option<Version>> {
+        Self::find_tree_root_at_or_before(state_merkle_db, u64::max_value())
+    }
+
+    fn find_tree_root_at_or_before(
+        state_merkle_db: &DB,
+        version: Version,
+    ) -> Result<Option<Version>> {
+        let mut iter = state_merkle_db.rev_iter::<JellyfishMerkleNodeSchema>(Default::default())?;
+        iter.seek_for_prev(&NodeKey::new_empty_path(version))?;
+        Ok(iter.next().transpose()?.map(|item| item.0.version()))
```

Comment:
Need to confirm root exists -- it can be a version pruned by the intra-epoch pruner.. 

oh.. the `get_state_snapshot_before()` call on the db side probably needs fix too >_<

Reply:
Done.

Will change the get_state_snapshot_before() on db side in a separate PR.